### PR TITLE
mgr/volumes: improve mgr/ssh support for mds deployment via 'fs volume create'

### PR DESCRIPTION
**Pull request by jschmid1**
The mds vstart deployment is currently broken for the ssh-orchestrator.

vstart first spawns an MDS daemon and then tries to create a fs

`ceph fs volume create a`

This calls in the mgr/orch backend to also create a MDS daemon. This will work for rook at it doesn't need an explicit host. The ssh-orch however does need it.

So I added parameter for `host` (optional)

old:
ceph fs volume create <fs_name>

new:
ceph fs volume create <fs_name> (hostname)

This will allow the ssh-orchestrator to deploy
a MDS daemon on the provided host.
If this is not specified and the ssh-backend is used
it will result in an RuntimeError. The volumes will
still be created though. 

(still reports a failure but doesn't break vstart)

edit: 

volume removal seems to be broken as it doesn't remove the daemon correctly. I'll follow up on this in another PR. (see https://github.com/ceph/ceph/pull/31762)

Signed-off-by: Joshua Schmid <jschmid@suse.de>

### CODE REVIEW

File: src/pybind/mgr/volumes/fs/volume.py [hunk 1]
```diff
@@ -272,24 +272,27 @@ def remove_filesystem(self, fs_name, confirm):
         command = {'prefix': 'fs rm', 'fs_name': fs_name, 'yes_i_really_mean_it': True}
         return self.mgr.mon_command(command)
 
-    def create_mds(self, fs_name):
-        spec = orchestrator.StatelessServiceSpec(fs_name)
+    def create_mds(self, fs_name, host):
+        spec = orchestrator.StatelessServiceSpec(fs_name,
+            placement=orchestrator.PlacementSpec(nodes=[host]))
```
Comment (sebastian-philipp) [inline src/pybind/mgr/volumes/fs/volume.py]: By default, `host` is an empty string, see `_cmd_fs_volume_create`:

```
$  . .tox/py3/bin/activate
$  UNITTEST=true python 
Python 3.6.8 (default, Aug 20 2019, 17:12:48) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import tests
>>> import orchestrator
>>> orchestrator.PlacementSpec(nodes=['']).nodes
[HostSpec(hostname='', network='', name='')]
```

Having an empty `HostSpec` doesn't really make sense. Can we make it an empty list, if host is empty?

**Comment by sebastian-philipp** [comment]
Could you prefix your PR with `mgr/volumes`?

File: src/pybind/mgr/orchestrator.py [hunk 2]
```diff
@@ -670,7 +670,8 @@ class PlacementSpec(object):
     def __init__(self, label=None, nodes=[]):
         self.label = label
 
-        self.nodes = [parse_host_specs(x, require_network=False) for x in nodes]
+        self.nodes = [parse_host_specs(x, require_network=False) for x in nodes if x]
+
```
Comment (sebastian-philipp) [inline src/pybind/mgr/orchestrator.py]: remove this empty line please. (PEP8)

File: src/pybind/mgr/ssh/module.py [hunk 3]
```diff
@@ -1036,7 +1036,7 @@ def update_mgrs(self, num, host_specs):
 
     def add_mds(self, spec):
         if not spec.placement.nodes or len(spec.placement.nodes) < spec.count:
-            raise RuntimeError("must specify at least %d hosts" % spec.count)
+            raise RuntimeError("Must specify at least %d hosts when using the SSH-Orchestrator." % spec.count)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/ssh/module.py]:
```suggestion
            raise OrchestratorValidationError("Must specify at least %d hosts when using the SSH-Orchestrator." % spec.count)
```

File: src/pybind/mgr/volumes/fs/volume.py [hunk 4]
```diff
@@ -272,24 +272,27 @@ def remove_filesystem(self, fs_name, confirm):
...
             # bubble out to the user
             log.exception("Failed to create MDS daemons")
             return -errno.EINVAL, "", str(e)
-        return 0, "", ""
+        return 0, "", "Created MDS daemon"
 
     ### volume operations -- create, rm, ls
 
-    def create_volume(self, volname):
+    def create_volume(self, volname, host):
```
Comment (sebastian-philipp) [inline src/pybind/mgr/volumes/fs/volume.py]: can we rename this parameter to `mds_host`?

**Comment by liewegas** [comment]
I don't think we should let the host requirement bubble up into other interfaces.  Instead, we should teach the orchestrator how to place daemons..either via node labels or a simple scheduler.

**Comment by tserong** [comment]
> I don't think we should let the host requirement bubble up into other interfaces. Instead, we should teach the orchestrator how to place daemons..either via node labels or a simple scheduler.

If using node labels, that would mean bubbling up the label requirement into the other interfaces, wouldn't it? How's that especially different to bubbling up the host requirement, given PlacementSpec in the orchestrator API allows either form?

(I'm wary of implementing a scheduler, even a simple one: isn't that what rook is for?)

File: src/pybind/mgr/orchestrator.py (hunk 2 above)
Comment (jschmid1) [reply src/pybind/mgr/orchestrator.py]: PEP8 expects 2 blank lines after a class definition. The rest of the file is structured the same way.

File: src/pybind/mgr/volumes/fs/volume.py (hunk 4 above)
Comment (jschmid1) [reply src/pybind/mgr/volumes/fs/volume.py]: will do

**Comment by jschmid1** [comment]
> I don't think we should let the host requirement bubble up into other interfaces. Instead, we should teach the orchestrator how to place daemons..either via node labels or a simple scheduler.

Basically what @tserong said, plus:

Even if we add something like labels/schedulers/roles we probably need a way to overwrite those. Having an optional parameter that allows to specifically target a node is probably not _too_ bad (for testing maybe?).

**Comment by sebastian-philipp** [comment]
> > I don't think we should let the host requirement bubble up into other interfaces. Instead, we should teach the orchestrator how to place daemons..either via node labels or a simple scheduler.

We kind of have to choose between two approaches here:

1. Allow `ceph fs volume create` to cover everything that is needed in order to deploy mds daemons. Then inevitably things will bubble up here. (Maybe we can change it into a generic `--placement-spec` argument that would then also cover the mds `count` in the CR.)
2. Require two commands to set up a fs volume: 
    ```
    ceph orchestrator ...mds...foo...bar....  (<count>|<hosts>|<labels>) 
    ceph fs volume create
    ```
    Then things will not bubble up in `fs volume create`, but then we don't *need* an orchestrator 
    integration in `mgr/volumes` at all. Instead we can just deploy the daemons in the orchestrator 
    command. 

I personally prefet to be able to have `fs volume create` be the only command needed.

> 
> If using node labels, that would mean bubbling up the label requirement into the other interfaces, wouldn't it? How's that especially different to bubbling up the host requirement, given PlacementSpec in the orchestrator API allows either form?

Yep. PlacementSpec will bubble up inevitably somewhere.

> 
> (I'm wary of implementing a scheduler, even a simple one: isn't that what rook is for?)

:+1:. I don't mind having a optional scheduler at some point, as long as it provides reasonable good results.

**Comment by liewegas** [comment]
I like the idea of a generate placement spec/string.  That way it can be hosts today, labels tomorrow, or nothing at all if ssh orch gets a scheduler.

**Comment by liewegas** [comment]
FWIW my preference would be to start with a trivial scheduler immediately (e.g., pick a node with the fewest daemons) so that everything "works", and then incrementally make it smarter (e.g. use labels with predictable names like 'mds', 'mds-$fsname', etc if they are present).  That way a well-labeled cluster will work even with no placement hints for these commands.

**Comment by jschmid1** [comment]
Shall we get this in and then focus on the label/roles stuff? This will at least fix the vstart mds creation

**Comment by batrick** [comment]
> > > I don't think we should let the host requirement bubble up into other interfaces. Instead, we should teach the orchestrator how to place daemons..either via node labels or a simple scheduler.
> 
> We kind of have to choose between two approaches here:
> 
>     1. Allow `ceph fs volume create` to cover everything that is needed in order to deploy mds daemons. Then inevitably things will bubble up here. (Maybe we can change it into a generic `--placement-spec` argument that would then also cover the mds `count` in the CR.)
> 
>     2. Require two commands to set up a fs volume:
>        ```
>        ceph orchestrator ...mds...foo...bar....  (<count>|<hosts>|<labels>) 
>        ceph fs volume create
>        ```
>        
>        
>        Then things will not bubble up in `fs volume create`, but then we don't _need_ an orchestrator
>        integration in `mgr/volumes` at all. Instead we can just deploy the daemons in the orchestrator
>        command.

Eventually there will be a separate module for monitoring the health of file systems and deploying MDS in response to a degraded file system (e.g. a new one). The volumes plugin currently calls into the orchestrator to deploy an MDS but this won't be the case in the future. For that reason, I don't want this "host" parameter added to the volumes plugin. See also: https://tracker.ceph.com/issues/40929

> I personally prefet to be able to have `fs volume create` be the only command needed.
> 
> > If using node labels, that would mean bubbling up the label requirement into the other interfaces, wouldn't it? How's that especially different to bubbling up the host requirement, given PlacementSpec in the orchestrator API allows either form?
> 
> Yep. PlacementSpec will bubble up inevitably somewhere.

I'm not familiar with PlacementSpec yet. Is there not a way to have a "default" behavior that places the MDS randomly on a host? Let that be the default for the ssh orchestrator in this case?

**Comment by liewegas** [comment]
Agree--let's plow ahead with labels.  And perhaps in the (very short) meantime, place (stateless) daemons on a random host.

**Comment by jschmid1** [comment]
I'll have a look into a very simple placement algorithm

**Comment by sebastian-philipp** [comment]
> I'll have a look into a very simple placement algorithm

For this specific PR, what about just accepting a general plcaement-spec that also supports adding labels? This way we would solve the immediate problem and support labels and properly support `count` for mgr/rook

**Comment by jschmid1** [comment]
since we how have basic labeling, I'll close this and integrate the still needed changes in a new PR.

**Pull request closed without merging**